### ISSUE TITLE
pandas dataframe in .uns causes write error

### ISSUE THREAD

**Issue by DanSchnell** [theislab/anndata#539]
Hi,
.uns seems to hold a pandas dataframe just fine, but adata.write seems to get hung up writing it out as part of the h5 file...
Is there a way to place a pandas dataframe in .uns so that adata.write will write it out?

**The dataframe types are:**
bdata.uns['Translation_table'].dtypes
information    object
optcum         object
seurat         object
scanpy         object
czi            object
dtype: object

**Error message is pasted in below:**

 bdata.write(filename=fn,compression='gzip')
Traceback (most recent call last):
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_io/utils.py", line 188, in func_wrapper
    return func(elem, key, val, *args, **kwargs)
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_io/h5ad.py", line 279, in write_series
    **dataset_kwargs,
  File "/Users/schh5o/.local/lib/python3.7/site-packages/h5py/_hl/group.py", line 148, in create_dataset
    dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
  File "/Users/schh5o/.local/lib/python3.7/site-packages/h5py/_hl/dataset.py", line 140, in make_new_dset
    dset_id.write(h5s.ALL, h5s.ALL, data)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5d.pyx", line 232, in h5py.h5d.DatasetID.write
  File "h5py/_proxy.pyx", line 145, in h5py._proxy.dset_rw
  File "h5py/_conv.pyx", line 444, in h5py._conv.str2vlen
  File "h5py/_conv.pyx", line 95, in h5py._conv.generic_converter
  File "h5py/_conv.pyx", line 249, in h5py._conv.conv_str2vlen
TypeError: Can't implicitly convert non-string objects to strings

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_io/utils.py", line 188, in func_wrapper
    return func(elem, key, val, *args, **kwargs)
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_io/h5ad.py", line 268, in write_dataframe
    write_series(group, colname, series, dataset_kwargs=dataset_kwargs)
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_io/utils.py", line 195, in func_wrapper
    ) from e
TypeError: Can't implicitly convert non-string objects to strings

Above error raised while writing key 'optcum' of <class 'h5py._hl.group.Group'> from /.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_core/anndata.py", line 1850, in write_h5ad
    as_dense=as_dense,
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_io/h5ad.py", line 125, in write_h5ad
    write_attribute(f, "uns", adata.uns, dataset_kwargs=dataset_kwargs)
  File "//anaconda3/lib/python3.7/functools.py", line 827, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_io/h5ad.py", line 139, in write_attribute_h5ad
    _write_method(type(value))(f, key, value, *args, **kwargs)
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_io/h5ad.py", line 299, in write_mapping
    write_attribute(f, f"{key}/{sub_key}", sub_value, dataset_kwargs=dataset_kwargs)
  File "//anaconda3/lib/python3.7/functools.py", line 827, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_io/h5ad.py", line 139, in write_attribute_h5ad
    _write_method(type(value))(f, key, value, *args, **kwargs)
  File "/Users/schh5o/.local/lib/python3.7/site-packages/anndata/_io/utils.py", line 195, in func_wrapper
    ) from e
TypeError: Can't implicitly convert non-string objects to strings

Above error raised while writing key 'optcum' of <class 'h5py._hl.group.Group'> from /.

Above error raised while writing key 'uns/Translation_table' of <class 'h5py._hl.files.File'> from /.
>>>

**Comment by ivirshup** [comment]
Can you report your version info and some information about what is in the dataframe?

This works on current release:

<details>
<summary> Example </summary>

```python
In [1]: import anndata

In [2]: import pandas as pd

In [3]: anndata.AnnData(uns={"df": pd.DataFrame({"a": [1, 2, 3]})}).write("test-uns-df.h5ad")

In [4]: anndata.read_h5ad("test-uns-df.h5ad")
Out[4]: 
AnnData object with n_obs × n_vars = 0 × 0
    uns: 'df'

In [5]: sinfo()
-----
IPython     7.21.0
anndata     0.7.5
pandas      1.2.2
sinfo       0.3.1
-----
Python 3.8.5 (default, Sep  4 2020, 02:22:02) [Clang 10.0.0 ]
macOS-10.15.7-x86_64-i386-64bit
16 logical CPU cores, i386
-----
Session information updated at 2021-04-07 11:54

In [8]: sinfo(dependencies=True)
-----
IPython     7.21.0
anndata     0.7.5
pandas      1.2.2
sinfo       0.3.1
-----
IPython             7.21.0
anndata             0.7.5
backcall            0.2.0
cython_runtime      NA
dateutil            2.8.1
decorator           4.4.2
h5py                3.1.0
ipython_genutils    0.2.0
jedi                0.17.2
mpl_toolkits        NA
natsort             7.1.1
numpy               1.20.1
packaging           20.9
pandas              1.2.2
parso               0.7.0
pexpect             4.8.0
pickleshare         0.7.5
prompt_toolkit      3.0.8
ptyprocess          0.7.0
pygments            2.8.1
pytz                2021.1
scipy               1.6.1
sinfo               0.3.1
six                 1.15.0
storemagic          NA
traitlets           5.0.5
wcwidth             0.2.5
-----
Python 3.8.5 (default, Sep  4 2020, 02:22:02) [Clang 10.0.0 ]
macOS-10.15.7-x86_64-i386-64bit
16 logical CPU cores, i386
-----
Session information updated at 2021-04-07 11:54
```
</details>

**Comment by DanSchnell** [comment]
Thanks for the example ivirshup. I believe that it does work...

I think the problem may have been due to either having some missing values in the df and/or the columns being of type "object"

I replaced the missing values with a character string and also created a modified df using:

TTable=Translation_table.astype("category")

With these 2 changes, the h5 write with TTable in .uns was successful.

Thanks again for your response.
